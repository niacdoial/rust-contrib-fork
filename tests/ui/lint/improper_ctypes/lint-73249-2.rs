#![feature(type_alias_impl_trait)]
#![deny(improper_ctypes)]

trait Baz {}

impl Baz for () {}

type Qux = impl Baz;

#[define_opaque(Qux)]
fn assign() -> Qux {}

trait Foo {
    type Assoc: 'static;
}

impl Foo for () {
    type Assoc = Qux;
}

#[repr(transparent)]
struct A<T: Foo> {
    x: &'static <T as Foo>::Assoc,
}

extern "C" {
    // possible FIXME: currenty, the error comes from the non-option'd reference, not the unsafety
    // of Qux
    fn lint_me() -> A<()>; //~ ERROR: uses type `A<()>`
}

fn main() {}
