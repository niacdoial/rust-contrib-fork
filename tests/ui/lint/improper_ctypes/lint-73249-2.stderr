error: `extern` block uses type `A<()>`, which is not FFI-safe
  --> $DIR/lint-73249-2.rs:29:21
   |
LL |     fn lint_me() -> A<()>;
   |                     ^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`A<()>`) is FFI-unsafe due to a `&Qux` field
note: the type is defined here
  --> $DIR/lint-73249-2.rs:22:1
   |
LL | struct A<T: Foo> {
   | ^^^^^^^^^^^^^^^^
   = help: consider using a raw pointer, or wrapping `&Qux` in an `Option<_>`
   = note: boxes, references, and function pointers are assumed to be valid (non-null, non-dangling, aligned) pointers,
           which cannot be garanteed if their values are produced by non-rust code
note: the lint level is defined here
  --> $DIR/lint-73249-2.rs:2:9
   |
LL | #![deny(improper_ctypes)]
   |         ^^^^^^^^^^^^^^^

error: aborting due to 1 previous error

